// telegraph/lib/websockets/index.js
// Telegraph Sockets
// Sets up websockets (currently through socket.io)
// -------------------------------------------------- //

module.exports.setup = function(server) {
    
    server.io = require('socket.io').listen(server.express);
    
    server.io.configure(function() {
        server.io.set("log level", (server.silent_mode) ? 0 : 2);
    });

    server.websocket_events = [];

    server.send = function(name, values) {
        server.io.sockets.emit(name, values);
    };

    server.volley = function(name, values) {
        server.io.sockets.volatile.emit(name, values);
    };

    server.broadcast = function(name, values) {
        server.io.sockets.broadcast.emit(name, values);
    };


    // Extend the "on" function to support websockets
    // -------------------------------------------------- //

    Telegraph.prototype.register  = Telegraph.prototype.on,

    Telegraph.on = function(name, action) {

        var self = this,
            event = name.split(":")[0],
            label = name.split(":")[1];
        
        if(event === "websocket") {

            self.register(name, action);
            
            self.websocket_events.push({
                name   : label,
                action : function(data) {
                    self.info("emitting " + name);
                    self.emit(name, data);
                }
            });

        } else {
            self.register(name, action);
        }

        return self;
    };

    
};


module.exports.install = function(server) {

    var events = server.websocket_events;
    
    server.io.sockets.on("connection", function(socket) {

        server.emit("websocket:connect", socket);

        // Add websocket events
        events.forEach(function(e) {

            // Is it the connection event?
            if (e.name === "connect") {
                // yes : do nothing
            } else { 
                // no : register the event
                socket.on(e.name, e.action);
            }
            
        });

    });

};
