// telegraph/lib/telegraph.js
// Telegraph Server
// Nate Hunzaker
//
// A framework based on Backbone for event driven applications


//-- Requirements ------------------------------------------------------//

// Side Dishes
require('colors');

var _            = require('underscore')
,   EventEmitter = require("events").EventEmitter
,   ejs          = require('ejs')
,   express      = require('express')
,   config       = require("konphyg")(process.cwd() + "/config")
,   util         = require('util')
;

module.exports = Telegraph = function (options) {
    
    //-- Defaults -------------------------------------------------------//

    var settings = {
        'appname'        : "Telegraph",
        'csrf'           : true,
        'routes'         : [],
        'websockets'     : false,
        'logger'         : require("./logger")
    },

    self = this;

    // Add Default Overides
    _.extend(self, settings, options);

    // Enable wiretaps
    require("./wiretap.js")(self);

    // Enable process listeners
    self._process_listeners();

    // Remaining setup
    self._setup();

};

Telegraph.prototype = new EventEmitter;

// Backbone Server Functionality
_.extend(Telegraph.prototype, {
    
    load: require("./helpers/utilities").load,
    
    //-- Reporting -----------------------------------------------------//

    log: function(type, strings) {
        var message = util.format.apply(this, strings);
        this.logger[type](message);
    },
    
    info: function() {
        this.log("info", arguments);
        return this;
    },
    
    error: function() {
        this.log("error", arguments);
        return this;
    },

    warn: function() {
        this.log("warn", arguments);
        return this;
    },

    success: function() {
        this.log("success", arguments);
        return this;
    },

    //-- Express Configuration Methods -------------------------------//

    use: function(middleware, config) {
        
        if (typeof middleware === "string") {
            middleware = express[middleware](config);            
        }

        this.express.use(middleware, config);

        return this;
    },
    
    set: function(name, config) {
        this.express.set(name, config);

        return this;
    },

    mount: function(route, app) {

        app = (typeof app === 'string') ? require(process.cwd() + "/" + app) : app;

        this.info("Mounting an application to %s", route);
        this.express.use(route, app);
        return this;
    },

    //-- Database Methods  --------------------------------------------//
    
    _db_connect: function() {

        var env = global.NODE_ENV || "development";

        this.database = config("database")[env];
        if ("database" in this) require("./database").setup(this);

    },

    //-- Server Methods ------------------------------------------------//

    _setup: function() {   

        var self = this,
            pdir = process.cwd() + "/public";
        
        var server = this.express = express.createServer(

            express.cookieParser(),
            express.session({ secret: require("./helpers/sessions").hash() }),
            
            express.bodyParser(),
            express.methodOverride(),

            express.favicon(),

            express.static(pdir),

            function(req, res) {
                // On overide for the index
                req.url = (req.url === "/") ? "index" : req.url;
                
                var event = req.method.toLowerCase() + ":" + req.url;
                self.emit(event, req, res);
                self.info("emitting " + event);
            }

        );

        // CSRF Management
        if (this.csrf) server.use(express.csrf());

        
        // Set view engine settings
        server.set("views", "app/views");
        server.set("view engine", "ejs");
        server.set("view options", {
            layout: process.cwd() + "/app/views/layout"
        });
        

        // Set Form Helpers
        require("./helpers/forms")(this);


        // Connect database
        this._db_connect();


        // Sockets
        if (this.websockets){
            require("./websockets").setup(this);
            require("./websockets").install(this);
        }

        return this;

    },

    start: function(port) {
        
        // Bulletproof the port
        port = process.env.PORT || port || 3000;

        // Load App Files
        this.load("app/models").load("app");

        this.express.listen(port);

        this.info(this.appname + " is listening on port " + port);
        
        return this;
    },
    
    _process_listeners: function() {

        var self = this;

        process.on("message", function(data) {

            switch(data.message) {
                
            case "start": 
                self.start(data.port); break;
                
            case "stop" :
                process.exit(); 
                break;
            }
        });
        
    }
    
});
